src/test/e2e-emulated/partial-eval.spec.ts

```diff
@@ -0,0 +1,209 @@
+import { ValueExpression } from "../../optimizer/types";
+
+const additiveExpressions = [
+    { original: "X + 3 + 1", simplified: "X + 4" },
+    { original: "3 + X + 1", simplified: "X + 4" },
+    { original: "1 + (X + 3)", simplified: "X + 4" },
+    { original: "1 + (3 + X)", simplified: "4 + X" },
+
+    // Should NOT simplify to X + 2, because X could be MAX - 2,
+    // so that X + 3 causes an overflow, but X + 2 does not overflow
+    { original: "X + 3 - 1", simplified: "X + 3 - 1" },
+    { original: "3 + X - 1", simplified: "3 + X - 1" },
+
+    // Should NOT simplify to X - 2, because X could be MIN + 2
+    { original: "1 + (X - 3)", simplified: "1 + (X - 3)" },
+
+    { original: "1 + (3 - X)", simplified: "4 - X" },
+
+    { original: "X + 3 - (-1)", simplified: "X + 4" },
+    { original: "3 + X - (-1)", simplified: "X + 4" },
+
+    // Should NOT simplify, because the current rules require that - commutes,
+    // which does not. This could be fixed in future rules.
+    { original: "-1 + (X - 3)", simplified: "-1 + (X - 3)" },
+
+    // Should NOT simplify to 2 - X, because X could be MIN + 3,
+    // so that 3 - X = -MIN = MAX + 1 causes an overflow,
+    // but 2 - X = -MIN - 1 = MAX does not
+    { original: "-1 + (3 - X)", simplified: "-1 + (3 - X)" },
+
+    // All the following cases should NOT simplify because -
+    // does not associate on the left with - or +.
+    // The following "associative rule" for - will be added in the future:
+    // (x - c1) op c2 -----> x + (-c1 op c2), where op \in {-,+}
+    { original: "1 - (X + 3)", simplified: "1 - (X + 3)" },
+    { original: "1 - (3 + X)", simplified: "1 - (3 + X)" },
+    { original: "1 - X + 3", simplified: "1 - X + 3" },
+    { original: "X - 1 + 3", simplified: "X - 1 + 3" },
+    { original: "1 - (X - 3)", simplified: "1 - (X - 3)" },
+    { original: "1 - (3 - X)", simplified: "1 - (3 - X)" },
+    { original: "1 - X - 3", simplified: "1 - X - 3" },
+    { original: "X - 1 - 3", simplified: "X - 1 - 3" },
+];
+
+const multiplicativeExpressions = [
+    { original: "X * 3 * 2", simplified: "X * 6" },
+    { original: "3 * X * 2", simplified: "X * 6" },
+    { original: "2 * (X * 3)", simplified: "X * 6" },
+    { original: "2 * (3 * X)", simplified: "6 * X" },
+
+    { original: "X * -3 * -2", simplified: "X * 6" },
+    { original: "-3 * X * -2", simplified: "X * 6" },
+    { original: "-2 * (X * -3)", simplified: "X * 6" },
+    { original: "-2 * (-3 * X)", simplified: "6 * X" },
+
+    // The following 4 cases should NOT simplify to X * 0.
+    // the reason is that X could be MAX, so that X*3 causes
+    // an overflow, but X*0 does not.
+    { original: "X * 3 * 0", simplified: "X * 3 * 0" },
+    { original: "3 * X * 0", simplified: "3 * X * 0" },
+    { original: "0 * (X * 3)", simplified: "0 * (X * 3)" },
+    { original: "0 * (3 * X)", simplified: "0 * (3 * X)" },
+
+    { original: "X * 0 * 3", simplified: "X * 0" },
```

Comment:
wondering why this is not `simplified: 0`, is it because the corresponding rule is not implemented yet?